src/gpio.rs

```diff
@@ -1373,6 +1376,159 @@ macro_rules! pin {
         impl_touch!($pxi: $pin, $touch: $touchno);
     };
 }
+#[cfg(feature = "nightly")]
+impl<T: Pin, MODE: InputMode> PinDriver<'_, T, MODE> {
+    async fn wait_for_high(&mut self) -> Result<(), EspError> {
+        InputFuture::new(self, InterruptType::HighLevel)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_low(&mut self) -> Result<(), EspError> {
+        InputFuture::new(self, InterruptType::LowLevel)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_rising_edge(&mut self) -> Result<(), EspError> {
+        InputFuture::new(self, InterruptType::PosEdge)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_falling_edge(&mut self) -> Result<(), EspError> {
+        InputFuture::new(self, InterruptType::NegEdge)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_any_edge(&mut self) -> Result<(), EspError> {
+        InputFuture::new(self, InterruptType::AnyEdge)?.await;
+        Ok(())
+    }
+}
+#[cfg(feature = "nightly")]
```

Comment:
No need for nightly, but does need `riscv-ulp-hal`, because the async code cannot run on the ULP, and the ULP does not have the `core::sync::atomic` stuff.